core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -636,21 +545,18 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        if (frozen && usePreComputedHashCode) {
```

Comment:
The hash code is rarely used on the data nodes (which is why I prefer to calculate it lazily). Now that ShardRouting is immutable we can use the racy single-check idiom (which is used in the JDK to calculate `java.lang.String#hashCode()`) without using volatile reads.